Add VPC endpoints

jsf9k opened the pull request:
## 🗣 Description ##

This pull request:
* Adds several VPC endpoints for various AWS services
* Configures the operations and private subnets to make use of said VPC endpoints
* Shares a single VPC route table among the private subnets.  Since the private subnets do not each use a different NAT gateway in a different operations subnet, there is no need for them to have separate VPC route tables.

(It would be nice to remove the NAT gateway from the operations subnet, since the private subnets no longer require it to communicate with AWS services, but they still require it to pull the Docker images from Docker Hub for the Guacamole Docker composition.)

## 💭 Motivation and context ##

Resolves #94.

## 🧪 Testing ##

I deployed these changes to env0 in our staging COOL environment and verified that the cloud-init scripts, CloudWatch agent, and SSM agent are all running without errors on all private and operations instances.  I also verified that the FreeIPA initialization for the guacamole instance still works.

I also redeployed with the external route removed from the operations subnet's route table and verified that everything still worked, with the exception of the downloading of the Docker images required for the Guacamole Docker composition.  This serves as an ironclad check that all AWS API calls are being made to the VPC endpoints and none are being made to the public endpoints.

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] _All_ future TODOs are captured in issues, which are referenced in code comments.
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All new and existing tests pass.

jsf9k committed:
Remove NAT gateway and associated route

jsf9k committed:
Use a single route table for all private subnets

Now that we are no longer using a separate NAT gateway for each
private subnet, the private subnets' route tables are all the same.

jsf9k committed:
Add permissions for creating, modifying, and deleting VPC endpoints

jsf9k committed:
Add security groups for instances using VPC endpoints to access the STS and SSM services

jsf9k committed:
Add VPC endpoints requires to use the STS, SSM, and S3 services

jsf9k committed:
Add a route for S3 access via the S3 VPC gateway endpoint

jsf9k committed:
Modify desktop gateway SG rules to use VPC endpoints

* Remove egress via 443 to 0.0.0.0/0
* Allow egress via 443 to the STS and SSM security groups
* Allow egress via 443 to the S3 gateway endpoint

jsf9k committed:
Add some CloudWatch VPC interface endpoints

These are necessary for the CloudWatch Agent to work.  This agent is responsible for forwarding our system logs to CloudWatch.

jsf9k committed:
Allow operations subnet to use the VPC endpoints in private subnet

This is necessary because adding the VPC endpoints to the VPCs makes
the VPC-internal DNS for sts.us-east-1.awsamazon.com, for example,
resolve to the internal IP of the VPC endpoint instead of a public AWS
IP.

jsf9k committed:
Add a route so the operations subnet can use the S3 gateway endpoint

The operations subnet could still access S3 via the internet gateway,
but it may as well use the gateway endpoint since it has to use the
interface endpoints for all the other AWS services.

Also rename some route-related Terraform resources to avoid collision
between operations resources and private resources.

jsf9k committed:
Revert "Remove NAT gateway and associated route"

This reverts commit 6b091f01e9fe8427a9a5fc9a4df3d0152af4762f.

I realized that the guacamole instance still needs the NAT gateway and
external web access in order to download the Docker images used in the
guacamole docker composition.

jsf9k committed:
Add SG and ACL rules to allow guacamole to use the NAT GW for HTTPS

This is necessary so that the guacamole instance can download the
Docker images required for the guacamole Docker composition.

dav3r commented:
> * Removes the NAT gateway from the operations subnet, since the private subnets no longer requires it to communicate with AWS services.

The description of this PR should be amended to remove this bullet, due to https://github.com/cisagov/cool-assessment-terraform/pull/93/commits/597a54164eb3dfbdd6e46e0bc1c8c70c0d50e46f. 

Also, doesn't this mean that https://github.com/cisagov/cool-assessment-terraform/pull/93/commits/a16b36a9756af5314b904c7a1e64bb4532b15d8f needs to be reverted as well?

jsf9k commented:
> > * Removes the NAT gateway from the operations subnet, since the private subnets no longer requires it to communicate with AWS services.
> 
> The description of this PR should be amended to remove this bullet, due to [597a541](https://github.com/cisagov/cool-assessment-terraform/commit/597a54164eb3dfbdd6e46e0bc1c8c70c0d50e46f).
> 
> Also, doesn't this mean that [a16b36a](https://github.com/cisagov/cool-assessment-terraform/commit/a16b36a9756af5314b904c7a1e64bb4532b15d8f) needs to be reverted as well?

I amended the PR description.

The commit a16b36a _does not_ need to be reverted.  As it turns out, the assessment VPCs were never set up like the shared services VPC with `N` public and `N` private subnets spread across AZs.  In that case we have each private subnet route its external traffic to the public subnet in the same AZ, so it makes sense to have separate route tables for each private subnet.  In this case the idea of separate route tables was implemented but not the more important idea of cross-AZ redundancy.

Sharp eyes!

jsf9k committed:
Improve some ACL rule comments

jsf9k committed:
Allow the operations subnet to use the STS VPC endpoint

jsf9k committed:
Specify the region name and region-specific endpoint URL for boto3 STS calls

STS used to be un-regioned, like S3, but now it is regioned.  This is
the one case where boto3 _does not_ do the right thing when you set
the region.  We have to set the region-specific endpint URL manually.

This is important since the STS VPC endpoint _only_ sets a local DNS
record to override the _local region's_ public STS endpoint.  If we
don't do this then boto3 will reach out to the _global_
https://sts.amazonaws.com URL, and that DNS entry will still point to
an external IP.

See this boto/boto3#1859 for more information about boto3's perverse
behavior in the case of STS.

jsf9k committed:
Specify the region name and region-specific endpoint URL for AWS CLI STS calls

STS used to be un-regioned, like S3, but now it is regioned.  The AWS
CLI by default uses the global endpoint URL https://sts.amazonaws.com
instead of the region-specific one, even when you specify the region.
Therefore, we have to set the region-specific endpint URL manually.

This is important since the STS VPC endpoint _only_ sets a local DNS
record to override the _local region's_ public STS endpoint.  If we
don't specify the endpoint URL then the AWS CLI will reach out to the
_global_ https://sts.amazonaws.com URL, and that DNS entry will still
point to an external IP.

dav3r commented on private_routing.tf:
```diff
@@ -12,46 +12,51 @@
+# we would assign each subnet its own NAT gateway, which would require
+# separate routing tables, but in this case we do not create any NAT
+# gateways since nothing in the private subnets requires access to the
+# internet outside the VPC.
```
This comment should be updated- the Guacamole instance in the private subnet still requires internet access to download the Guacamole Docker images and we still have the single NAT gateway used by the private subnets.

jsf9k replied:
See commit 5a40a3a.

hillaryj commented on sts_endpoint_sg_rules.tf:
```diff
@@ -0,0 +1,27 @@
+#
+# TODO - Only the Nessus instances require this access.  We could
+# avoid giving it to the other instance types by creating a separate
+# security group for Nessus instances.  See #95 for more details.
```
Do we want to include the link or just the number? It seems like we usually include a full link.

```suggestion
# security group for Nessus instances.  See #95 for more details:
# https://github.com/cisagov/cool-assessment-terraform/issues/95
```

hillaryj commented on vpc_endpoints.tf:
```diff
@@ -0,0 +1,226 @@
+
+  # Normally we would use a for_each here, but the private subnets are
+  # _all in the same AZ_ in this case.  Why even have multiple subnets
+  # if you're not going to spread them around?  Harumph!
```
Harumph! https://www.youtube.com/watch?v=LmnaeU0k80I

jsf9k replied:
Also this bit from "Blazing Saddles":
[![Harumph](https://img.youtube.com/vi/Yt3GBlVjUd0/0.jpg)](https://www.youtube.com/watch?v=Yt3GBlVjUd0)

hillaryj reviewed (approved):
One suggestion which is only a comment change, and otherwise: 🏆

dav3r reviewed (approved):
Excellent work and really great job on the comments!  🐎 

...except for one comment that needs a slight revision.

jsf9k committed:
Consistently name AWS region variable

I noticed that previously we were using the name "aws_region" in the
cloud-init templates, whereas I had used the name "region" in the code
that I added.  For consistency's sake, I changed the name in the code
that I added.

jsf9k committed:
Add issue link to a comment

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

jsf9k committed:
Update comment to agree with reality

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

felddy commented on private_acl_rules.tf:
```diff
@@ -139,7 +139,7 @@ resource "aws_network_acl_rule" "private_egress_to_operations_via_vnc" {
   egress         = true
   protocol       = "tcp"
-  rule_number    = 120 + index(var.private_subnet_cidr_blocks, each.value)
+  rule_number    = 130 + index(var.private_subnet_cidr_blocks, each.value)
```
🧙 Seems magical.  Is this counting by 10 jazz useful?  Does this break if index ≥ 10?

jsf9k replied:
Yes, it breaks if `index >= 10`.  Our numbering of the ACL rules has always sucked.  Do you have any thoughts as to how it can be improved?

felddy commented on subnets.tf:
```diff
@@ -45,3 +45,22 @@ resource "aws_internet_gateway" "assessment" {
+  vpc  = true
+}
+
+resource "aws_nat_gateway" "nat_gw" {
```
Should we think about including the Guacamole Docker images as part of the AMI so that we can re-remove this gateway?

jsf9k replied:
That is not a bad idea, but I have a sense of deja vu.  I think we discussed this before but decided against it, maybe because this was back when we were first cloudifying CyHy and making our AMIs public, and hence we wanted to keep sensitive data out of them.

@dav3r and @felddy, do you remember when we discussed this before?

dav3r replied:
I don't remember the details of this discussion from the past.  I am definitely in favor of baking the Guacamole Docker image into our AMI if it means we can get rid of the NAT gateway and related port 443 access rules.

jsf9k replied:
OK.  I created cisagov/cool-system#158 for this change.

mcdonnnj commented on cloud-init/install-certificates.py:
```diff
@@ -24,7 +25,25 @@
+# https://sts.amazonaws.com URL, and that DNS entry will still point
+# to an external IP.
+#
+# See this link for more information about boto3's perverse behavior
```
> boto3's perverse behavior

🤣

felddy commented on cloud-init/render-guac-connection-sql-template.py:
```diff
@@ -31,7 +32,23 @@
 
 # Create STS client
-sts = boto3.client("sts")
+#
```
Good catch!  👁️ 💪

felddy reviewed (approved):
Wonderful.  This makes a lot of sense. 💯 
I had a question about preloading the Docker images into the Guac AMI so that we can remove the NAT gateway.  We should discuss that. 🐳 🥑 

> My motto was always to keep swinging. Whether I was in a slump or feeling badly or having trouble off the field, the only thing to do was keep swinging. --Hank Aaron ⚾

mcdonnnj commented on operations_acl_rules.tf:
```diff
@@ -30,6 +30,23 @@ resource "aws_network_acl_rule" "operations_ingress_from_private_via_vnc" {
 }
 
+# Allow ingress from the private subnets via port 443.  This is
+# necessary so that the guacamole instance can download the Docker
```
```suggestion
# necessary so that the Guacamole instance can download the Docker
```

mcdonnnj commented on operations_acl_rules.tf:
```diff
@@ -30,6 +30,23 @@ resource "aws_network_acl_rule" "operations_ingress_from_private_via_vnc" {
 
+# Allow ingress from the private subnets via port 443.  This is
+# necessary so that the guacamole instance can download the Docker
+# images used in the docker composition via the NAT gateway.
```
```suggestion
# images used in the Docker composition via the NAT gateway.
```

mcdonnnj commented on operations_acl_rules.tf:
```diff
@@ -150,6 +167,10 @@ resource "aws_network_acl_rule" "operations_ingress_from_anywhere_via_icmp" {
 # Allow egress to anywhere via any protocol and port
 # For: Assessment team operational use
+#
+# Note that this also covers the return traffic when the guacamole
```
```suggestion
# Note that this also covers the return traffic when the Guacamole
```

mcdonnnj commented on operations_acl_rules.tf:
```diff
@@ -150,6 +167,10 @@ resource "aws_network_acl_rule" "operations_ingress_from_anywhere_via_icmp" {
 # For: Assessment team operational use
+#
+# Note that this also covers the return traffic when the guacamole
+# instance downloads the Docker images used in the docker composition
```
```suggestion
# instance downloads the Docker images used in the Docker composition
```

mcdonnnj commented on private_acl_rules.tf:
```diff
@@ -15,7 +15,14 @@ resource "aws_network_acl_rule" "private_ingress_from_cool_via_https" {
+#
+# For: Guacamole fetches its SSL certificate via boto3 (which uses
+# HTTPS).  It also needs to download the Docker images used in the
+# guacamole Docker composition.
```
```suggestion
# Guacamole Docker composition.
```

mcdonnnj commented on private_acl_rules.tf:
```diff
@@ -15,7 +15,14 @@ resource "aws_network_acl_rule" "private_ingress_from_cool_via_https" {
 # Allow egress to anywhere via HTTPS
-# For: Guacamole fetches its SSL certificate via boto3 (which uses HTTPS)
+#
+# For: Guacamole fetches its SSL certificate via boto3 (which uses
```
This is now inconsistent with the changes made here:
https://github.com/cisagov/cool-assessment-terraform/blob/5a40a3a80152600df6cad2d489a47f27eae17711/desktop_gateway_sg_rules.tf#L30-L31
and here:
https://github.com/cisagov/cool-assessment-terraform/blob/5a40a3a80152600df6cad2d489a47f27eae17711/desktop_gateway_sg_rules.tf#L75-L76

jsf9k replied:
I'm not following how it is inconsistent.  Can you elaborate?

Remember that with VPC _gateway_ endpoints (used only with S3 and DynamoDB), the traffic leaves the subnet destined for _the real, external IP_ to which `s3.[region.]amazonaws.com` resolves but is directed to the VPC gateway endpoint by the route table.  Hence the subnet's ACL needs to allow HTTPS egress anywhere.

mcdonnnj replied:
Sorry, I just meant the comment language changed, and this comment does not match the change.
`Guacamole fetches its SSL certificate via boto3 (which uses HTTPS)` -> `Guacamole requires access to S3 in order to download its certificate.`

jsf9k replied:
See commit d9efdca.

mcdonnnj commented on vpc_endpoints.tf:
```diff
@@ -0,0 +1,226 @@
+}
+
+# S3 gateway endpoint
+resource "aws_vpc_endpoint" "s3" {
```
Is there a reason this isn't with the other `aws_vpc_endpoint` declarations?

jsf9k replied:
See commit 2d42f57.

mcdonnnj replied:
I thought that might have been why, but the comments make it explicit. Thanks!

mcdonnnj commented on desktop_gateway_sg_rules.tf:
```diff
@@ -24,6 +25,66 @@ resource "aws_security_group_rule" "desktop_gw_egress_to_anywhere_via_https" {
+}
+
 # Allow ingress from COOL Shared Services VPN server CIDR block
 # via port 443 (nginx/guacamole web)
```
```suggestion
# via port 443 (nginx/Guacamole web)
```

mcdonnnj reviewed (approved):
This is a pretty big lift and some strong work 💪💪💪
I had some suggestions/questions, but I was also curious if we have a standard on capitalization of protocols? We seem to use for example `https` and `HTTPS` without clear purpose.

jsf9k committed:
Capitalize "Guacamole" and "Docker" in comments

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k commented:
> I was also curious if we have a standard on capitalization of protocols? We seem to use for example `https` and `HTTPS` without clear purpose.

We do not have such a standard, but you are welcome to propose one.  Capitalization is probably the way to go.

jsf9k committed:
Add comments to clearly separate VPC interface and gateway endpoints

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented:
> 
> 
> > I was also curious if we have a standard on capitalization of protocols? We seem to use for example `https` and `HTTPS` without clear purpose.
> 
> We do not have such a standard, but you are welcome to propose one. Capitalization is probably the way to go.

If I'm proposing, then yes, I would vote for defaulting to capitalization for protocols.

jsf9k commented:
> > > I was also curious if we have a standard on capitalization of protocols? We seem to use for example `https` and `HTTPS` without clear purpose.
> > 
> > 
> > We do not have such a standard, but you are welcome to propose one. Capitalization is probably the way to go.
> 
> If I'm proposing, then yes, I would vote for defaulting to capitalization for protocols.

You could add something to [this long-running PR](https://github.com/cisagov/development-guide/pull/44).

jsf9k committed:
Clean up comment formatting

jsf9k committed:
Reword comment for consistency with similar comments made elsewhere

Co-authored-by:  Nick M <50747025+mcdonnnj@users.noreply.github.com>
